pkg/util/cgroups/stats.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux
+// +build linux
+
+package cgroups
+
+// PSIStats represent the Pressure Stall Information data
+// Source:
+// cgroupv1: not present
+// cgroupv2: *.pressure
+type PSIStats struct {
+	Avg10  *float64 // Percentage (0-100)
+	Avg60  *float64 // Percentage (0-100)
+	Avg300 *float64 // Percentage (0-100)
+	Total  *uint64  // Nanoseconds
+}
+
+// MemoryStats - all metrics in bytes except if otherwise specified
+// All statistics are hierarchical (i.e. includes all descendants)
+// Source:
+// cgroupv1: memory controller
+// cgroupv2: memory controller
+type MemoryStats struct {
+	UsageTotal   *uint64
+	Cache        *uint64
+	Swap         *uint64
+	RSS          *uint64
+	RSSHuge      *uint64
+	MappedFile   *uint64
+	Pgpgin       *uint64 // Number (no unit), cgroupv1 only
+	Pgpgout      *uint64 // Number (no unit), cgroupv1 only
+	Pgfault      *uint64 // Number (no unit)
+	Pgmajfault   *uint64 // Number (no unit)
+	InactiveAnon *uint64
+	ActiveAnon   *uint64
+	InactiveFile *uint64
+	ActiveFile   *uint64
+	Unevictable  *uint64
+
+	KernelMemory *uint64
+
+	// This field is mapped to `memory.failcnt` for cgroupv1 and to "oom" in `memory.event`, it does not mean an OOMKill event happened.
+	OOMEvents     *uint64 // Number (no unit).
+	OOMKiilEvents *uint64 // cgroupv2 only
+
+	Limit             *uint64
+	MinThreshold      *uint64 // cgroupv2 only
+	LowThreshold      *uint64 // cgroupv1: mapped to soft_limit
+	HighThreshold     *uint64 // cgroupv2 only
+	SwapLimit         *uint64
+	SwapHighThreshold *uint64 // cgroupv2 only
+
+	PSISome PSIStats
+	PSIFull PSIStats
+}
+
+// CPUStats - all metrics are in nanoseconds execept if otherwise specified
+// cgroupv1: cpu/cpuacct/cpuset
+// cgroupv2: cpu/cpuset
+type CPUStats struct {
+	User   *uint64
+	System *uint64
+	Total  *uint64
+
+	Shares           *uint64 // Raw share value (no unit)
+	ElapsedPeriods   *uint64 // Number (no unit)
+	ThrottledPeriods *uint64 // Number (no unit)
+	ThrottledTime    *uint64
+	CPUCount         *uint64 // Number of accessible logical CPU (from cpuset.cpus)
+
+	SchedulerPeriod *uint64
+	SchedulerQuota  *uint64
+
+	PSISome PSIStats
+}
+
+// PIDStats store stats about running threads and processes
+// cgroupv1: pid
+// cgroupv2: pid
+type PIDStats struct {
+	PIDs                    []int
+	HierarchicalThreadCount *uint64 // Number of threads in cgroups + all children
+	HierarchicalThreadLimit *uint64 // Maximum number of threads in cgroups + all children
+}
+
+// DeviceIOStats is a sub-structure to store stats per device
+type DeviceIOStats struct {
+	ReadBytes       *uint64
+	WriteBytes      *uint64
+	ReadOperations  *uint64
+	WriteOperations *uint64
+
+	ReadBytesLimit       *uint64 // cgroupv2 only (bytes/s)
+	WriteBytesLimit      *uint64 // cgroupv2 only (bytes/s)
+	ReadOperationsLimit  *uint64 // cgroupv2 only (ops/s)
+	WriteOperationsLimit *uint64 // cgroupv2 only (ops/s)
```

Comment:
Any reason not to split into two different structs? Also the choice with using pointers - wouldn't generally the entire set of stats be missing if you fail to read a file? (having pointer to a struct as opposed to individual fields)

Reply:
I thought it's simpler to only have a single `map` holding all stats for a device. It's true that excluding parsing bugs or Kernel bugs, we should get all or nothing (per cgroup file). In this case coming from two files (`io.stat` and `io.max`).
However, I kept pointers everywhere for two reasons:
- Consistency
- You don't have to wonder if `0` is a legit value or not. For instance, for limits, only one of them can be set but setting limit to `0` is a legit value (preventing access to a device)